src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -4385,71 +4387,132 @@ private ParameterSyntax ParseParameter()
                 }
 
-                if (this.CurrentToken.Kind == SyntaxKind.EqualsToken)
-                {
-                    equals = this.EatToken(SyntaxKind.EqualsToken);
-                }
-                if (exclamationExclamation != null)
-                {
-                    exclamationExclamation = (exclamationExclamation.Kind != SyntaxKind.ExclamationExclamationToken)
-                        ? ConvertToMissingWithTrailingTrivia(exclamationExclamation, SyntaxKind.ExclamationExclamationToken)
-                        : CheckFeatureAvailability(exclamationExclamation, MessageID.IDS_ParameterNullChecking);
-                }
-                EqualsValueClauseSyntax def = null;
-                if (!(equals is null))
+                var type = this.ParseType(mode: ParseTypeMode.Parameter);
+                var identifier = this.ParseIdentifierToken();
+
+                // When the user type "int goo[]", give them a useful error
+                if (this.CurrentToken.Kind is SyntaxKind.OpenBracketToken && this.PeekToken(1).Kind is SyntaxKind.CloseBracketToken)
                 {
-                    var value = this.ParseExpressionCore();
-                    def = _syntaxFactory.EqualsValueClause(equals, value: value);
-                    def = CheckFeatureAvailability(def, MessageID.IDS_FeatureOptionalParameter);
+                    identifier = AddTrailingSkippedSyntax(identifier, SyntaxList.List(
+                        this.AddError(this.EatToken(), ErrorCode.ERR_BadArraySyntax),
+                        this.EatToken()));
                 }
-                return _syntaxFactory.Parameter(attributes, modifiers.ToList(), type, name, exclamationExclamation, def);
+
+                ParseParameterNullCheck(out var exclamationExclamationToken, out var equalsToken);
+
+                // If we didn't already consume an equals sign as part of !!=, then try to scan one out now.
+                equalsToken ??= TryEatToken(SyntaxKind.EqualsToken);
+
+                EqualsValueClauseSyntax equalsValueClause = null;
+                if (equalsToken != null)
+                    equalsValueClause = CheckFeatureAvailability(_syntaxFactory.EqualsValueClause(equalsToken, this.ParseExpressionCore()), MessageID.IDS_FeatureOptionalParameter);
+
+                return _syntaxFactory.Parameter(attributes, modifiers.ToList(), type, identifier, exclamationExclamationToken, equalsValueClause);
             }
             finally
             {
                 _pool.Free(modifiers);
             }
         }
 
+#nullable enable
+
+        /// <summary>
+        /// Parses out the <c>!!</c> following once the type and identifier has been parsed out already.  If the token
+        /// is followed by <c>!!=</c> or <c>! !=</c>, then the final equals will be returned out through <paramref
```

Comment:
The comment wording seems a little complicated. Perhaps:

```suggestion
        /// Parses the <c>!!</c> following a parameter type and identifier.  If the token
        /// is followed by <c>!!=</c> or <c>! !=</c>, then the final equals will be returned through <paramref
```